### DISCUSSION TITLE
Lesson 12: liquidate() - why do we only decrease amount of minted token of liquidated user in DSCEngine and not in DSC?

### DISCUSSION THREAD

**Discussion by Himchenhim** [Cyfrin/foundry-full-course-cu#1411] [Q&A] [answered]
Hello everyone! 

First of all, I want to thank for super froggy course from @PatrickAlphaC . It definitely the one, of the best course that I have ever participated in, because I have learnt A LOT!

Additionally, could I kindly ask you to help me with understanding `liquidate()` function?

some definitions:
for simplicity I consider collateralization only by ETH: 
deposited $ETH --- loan $DSC 
$DSC - dollars in DSC 
$ETH - dollars in ETH  

**Question**:
Do we as _liquidators_ need to pay somebody debt in **$DSC** or in **$ETH**?  + why do we only decrease amount of minted token in DSCEngine and not in DSC?
as I am reading the code, I think, that we need to pay it in  $DSC. 
Why?
```javascript
    function liquidate(address tokenCollateralAddress, address userToLiquidate, uint256 debtToCover)
        external
        moreThanZero(debtToCover)
        nonReentrant
    {
        
        uint256 startingUserHealthFactor = _healthFactor(userToLiquidate);
        if (startingUserHealthFactor >= MIN_HEALTH_FACTOR) {
            revert DSCEngine__HealthFactorOk();
        }
        
        // how much do we want to cover?
        uint256 tokenAmountFromDebtCovered = getTokenAmountFromUsd(tokenCollateralAddress, debtToCover);
        // how big bonus will we receive 
        uint256 bonusCollateral = (tokenAmountFromDebtCovered * LIQUIDATION_BONUS) / LIQUIDATION_PRECISION;
        
        // we are receiving: $ETH from collateral of **userToLiquidate** account + bonus also from **userToLiquidate** account 
        _redeemCollateral(
            tokenCollateralAddress, tokenAmountFromDebtCovered + bonusCollateral,  userToLiquidate, msg.sender
        );

        // the most confusing part for me:
 ---> _burnDsc(debtToCover, userToLiquidate, msg.sender);

        // checking if we help to  improve health factor
        uint256 endingUserHealthFactor = _healthFactor(userToLiquidate);
        if (endingUserHealthFactor <= MIN_HEALTH_FACTOR) {
            revert DSCEngine__HealthFactorNotImproved();
        }
        // just for sure
        _revertIfHealthFactorIsBroken(msg.sender);
    }
```
let's dive into `_burnDsc()`, where we PAY LOAN for "userToLiquidate" and _burn_ our $DSC

```javascript
    function _burnDsc(uint256 amount, address onBehalfOf, address dscFrom) private moreThanZero(amount) {
         // why we burn it only in DSCEngine and not burn $DSC of "userToLiquidate" using the same logic as burned $DSC of liquidators?
?->   // as I see usage of it's protocol, people **will take loan of $DSC** and can **use**(for paying) them for interacting with
        // another contract, which has some logic connected with $DSC(DecentralizedStableCoin). If I am not riqht, please correct me.   
        s_DSCMinted[onBehalfOf] -= amount;
        // burning $DSC of liquidators, BUT! Not burning them by
        // s_DSCMinted[dscFrom] -= amount 
        bool transferSuccess = i_dsc.transferFrom(dscFrom, address(this), amount);
        if (!transferSuccess) {
            revert DSCEngine__TransferFailed();
        }
        i_dsc.burn(amount);
    }
``` 

I am appreciating every answer and every help! :)

**Comment by PatrickAlphaC** [comment]
Great questions! 

> Do we as liquidators need to pay somebody debt in $DSC or in $ETH? + why do we only decrease amount of minted token in DSCEngine and not in DSC?

We have to _pay back_ the debt in $DSC. So we burn our own DSC to remove the debt of those users. We essentially "take on" their debt. You always pay back your debt with DSC. 

> as I am reading the code, I think, that we need to pay it in $DSC.

Correct! 

> Why?

You can think of minting DSC like a loan. When a user takes out a loan, they need to pay it back if they are undercollateralized (let's say the value of their collateral goes under), if they don't those users should be punished!

The way we do punishments in the DSC system, is that liquidators get to pay back _a fraction of the debt owned as a discount_ and in return they are awarded the underlying collateral.

For example:
1. User A: Mints 100 $DSC with 200 $ETH ($100 of DSC with $200 of ETH). User A has a healthy health factor, since their debt / collateral ratio is 200%
2. ETH price crashes, the original $200 of ETH is now worth $125.
3. The health factor of User A is now unhealthy with a ratio of 125%, and the minimum ratio is 150% (or whatever the min ratio is)! 
4. Liquidator B comes along, and pays back the $100 DSC debt, and get to take the entire $125 ETH collateral in return! 

So liquidator B did the following:
- Removed the "bad debt" from the system (User A being undercollateralized)
- Got User A's collateral at a discount since User A was undercollateralized!

**Reply by PatrickAlphaC** [reply]
Does that make sense?

**Reply by Himchenhim** [reply]
Thank you for the detailed answer, Patrick! I understand analogy with loan and it DOES make sense to me.

However, I don't understand, why we use this piece of code:
```javascript
s_DSCMinted[onBehalfOf] -= amount;
```
does it influence in anyway on `DecentralizedStableCoin`? For instance, if after execution of the line above, we will want to check balance of the user onBehalfOf using this call: 
```javascript
i_dsc.balanceOf(onBehalfOf);
```   
Will it return to us correct answer (the same as value of `s_DSCMinted[onBehalfOf]`)?

For example, what happens with liquidator B -> it receives collateral of user A + burn liquidator's B _DSC tokens_
Now, if we make the call:
 ```javascript
i_dsc.balanceOf(dscFrom);
``` 
we get correct answer (the same as value of `s_DSCMinted[dscFrom]`)

Please, feel free to correct me. I appreciate it!

**Reply by PatrickAlphaC** [reply, marked as answer]
> does it influence in anyway on DecentralizedStableCoin

Nope :) We never do a transfer. This is why we separated the COIN contract from the ENGINE contract, they each keep track of different things.

Coin contract (DSC.sol): Keeps track of real token balances
Engine contract (DSCEngine.sol): Keeps track of _owed_ token balances 

The `s_DSCMinted` mapping is a mapping that says `John owes us 7 coins, he should eventually have to pay us that back!". What's a little misleading is that the mapping is named `s_DSCMinted`, since Fred could have minted the tokens and John "token on Fred's debt".

Whereas the `balanceOf` keeps track of "John _has_ 7 coins".

**Reply by Himchenhim** [reply]
Wow! Okay, your insight is game changer 💥 for my understanding of this protocol. Thank you very much for helping to figure it out!

**Reply by adesokan247** [reply]
this really helped thanks alot for the question and answer

**Reply by EngrPips** [reply]
Indeed
